{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f05433d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import subprocess\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed\n",
    "\n",
    "\n",
    "location = \"/mnt/griffin/saubar/Species_sequences\" # location of the root folder where all data and analysis will be done \n",
    "\n",
    "\n",
    "list_of_fasta_files = os.listdir(f\"{location}/1.Species_sequences\") \n",
    "\n",
    "\n",
    "'''\n",
    "\"1.Species_sequences\" is the folder that contains species specific fasta files with sequences.\n",
    "This folder is present in the location i.e., inside Species_sequences folder.\n",
    "This contains files for all the species.\n",
    "Each fasta file has the following name Species.fas\n",
    "Each fasta file has the following structure:\n",
    ">BUSCOid1\n",
    "Sequence\n",
    ">BUSCOid2\n",
    "Sequences\n",
    "\n",
    "The BUSCO ids are shared between the species.\n",
    "\n",
    "'''\n",
    "print(\"Step 1: Preparing Sequences\")\n",
    "gene_dictionary = {}\n",
    "gene_name_list = []\n",
    "species_name_list = []\n",
    "for file_name in list_of_fasta_files:\n",
    "    species_name = file_name.split(\".\")[0]\n",
    "    if species_name not in species_name_list:\n",
    "        species_name_list.append(species_name)\n",
    "    gene_dictionary[species_name] = SeqIO.to_dict(SeqIO.parse(f\"{location}/1.Species_sequences/{file_name}\", \"fasta\"))\n",
    "    for gene_name in gene_dictionary[species_name]:\n",
    "        if gene_name not in gene_name_list:\n",
    "            gene_name_list.append(gene_name)\n",
    "\n",
    "list_of_folders =  os.listdir(f\"{location}\")\n",
    "if \"2.Gene_sequences\" not in list_of_folders:\n",
    "    os.mkdir(f\"{location}/2.Gene_sequences\")\n",
    "else:\n",
    "    subprocess.run(f\"rm -r {location}/2.Gene_sequences/*\", shell = True)\n",
    "\n",
    "for gene_name in gene_name_list:\n",
    "    output = ''\n",
    "    for species_name in species_name_list:\n",
    "        if gene_name in gene_dictionary[species_name]:\n",
    "            sequence = gene_dictionary[species_name][gene_name].seq\n",
    "            output += f\">{species_name}\\n{sequence}\\n\"\n",
    "    with open(f\"{location}/2.Gene_sequences/{gene_name}.fas\", 'w') as out_file:\n",
    "        out_file.write(output)\n",
    "print(\"Step 1: Done\")\n",
    "\n",
    "print(\"Step 2: Preparing Sequences for Mafft alignment\")\n",
    "\n",
    "list_of_fasta_files = os.listdir(f\"{location}/2.Gene_sequences/\")\n",
    "\n",
    "list_of_folders =  os.listdir(f\"{location}\")\n",
    "if \"3.Sequences_no_gap\" not in list_of_folders:\n",
    "    os.mkdir(f\"{location}/3.Sequences_no_gap\")\n",
    "else:\n",
    "    subprocess.run(f\"rm -r {location}/3.Sequences_no_gap/*\", shell = True)\n",
    "\n",
    "\n",
    "if \"4.Translated_sequences\" not in list_of_folders:\n",
    "    os.mkdir(f\"{location}/4.Translated_sequences\")\n",
    "else:\n",
    "    subprocess.run(f\"rm -r {location}/4.Translated_sequences/*\", shell = True)\n",
    "\n",
    "\n",
    "for i,fasta_file_name in enumerate(list_of_fasta_files):\n",
    "    if i % 100 == 0:\n",
    "        print(i,\"/\",len(list_of_fasta_files))\n",
    "    output = ''\n",
    "    output_nogap = ''\n",
    "    fasta_file = SeqIO.parse(f\"{location}/2.Gene_sequences/{fasta_file_name}\", 'fasta')\n",
    "    for records in fasta_file:\n",
    "        no_gap =  records.seq.replace(\"-\",'')\n",
    "        translated_sequence = no_gap.translate()\n",
    "        if translated_sequence[-1] == \"*\":\n",
    "            if \"*\" not in translated_sequence[:-1]:\n",
    "                output += f\">{records.id}\\n{translated_sequence[:-1]}\\n\"\n",
    "                output_nogap += f\">{records.id}\\n{no_gap}\\n\"\n",
    "        else:\n",
    "           if \"*\" not in translated_sequence:\n",
    "               output += f\">{records.id}\\n{translated_sequence}\\n\"\n",
    "               output_nogap += f\">{records.id}\\n{no_gap}\\n\"\n",
    "               \n",
    "    with open(f'{location}/4.Translated_sequences/{fasta_file_name}', 'w') as out_file:\n",
    "        out_file.write(output)\n",
    "    with open(f'{location}/3.Sequences_no_gap/{fasta_file_name}', 'w') as out_file:\n",
    "        out_file.write(output_nogap)\n",
    "\n",
    "\n",
    "print(\"Step 2: Done\")\n",
    "\n",
    "\n",
    "print(\"Step 3: Aligning\")\n",
    "list_of_folders =  os.listdir(f\"{location}\")\n",
    "if \"5.Aligned_MAFFT_protein\" not in list_of_folders:\n",
    "    os.mkdir(f\"{location}/5.Aligned_MAFFT_protein\")\n",
    "else:\n",
    "    subprocess.run(f\"rm -r {location}/5.Aligned_MAFFT_protein/*\", shell = True)\n",
    "\n",
    "list_of_fasta_files = os.listdir(f\"{location}/4.Translated_sequences\")\n",
    "list_of_done_files = os.listdir(f\"{location}/5.Aligned_MAFFT_protein\")\n",
    "\n",
    "\n",
    "\n",
    "def run_mafft(file_name):\n",
    "    cmd = (\n",
    "        f\"mafft --auto --thread 1 \"\n",
    "        f\"'{location}/4.Translated_sequences/{file_name}' > \"\n",
    "        f\"'{location}/5.Aligned_MAFFT_protein/{file_name}'\"\n",
    "    )\n",
    "    subprocess.run(cmd, shell=True)\n",
    "    return file_name\n",
    "\n",
    "with ThreadPoolExecutor(max_workers=10) as executor:\n",
    "    futures = [\n",
    "        executor.submit(run_mafft, f)\n",
    "        for f in list_of_fasta_files\n",
    "        if f not in list_of_done_files\n",
    "    ]\n",
    "\n",
    "    for i, future in enumerate(as_completed(futures), 1):\n",
    "        print(f\"{i}/{len(futures)} completed: {future.result()}\")\n",
    "\n",
    "\n",
    "def map_gaps_to_untranslated(translated_with_gaps, untranslated_no_gaps):\n",
    "    \"\"\"\n",
    "    Map the gaps in a translated sequence onto its corresponding untranslated sequence.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    translated_with_gaps : str\n",
    "        Translated sequence containing '-' for gaps (1 char per amino acid/codon).\n",
    "    untranslated_no_gaps : str\n",
    "        Untranslated nucleotide sequence without any gaps.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    str\n",
    "        Untranslated nucleotide sequence with gaps inserted to match the translated alignment.\n",
    "    \"\"\"\n",
    "    untranslated_with_gaps = []\n",
    "    nt_index = 0  # position in the nucleotide sequence\n",
    "\n",
    "    for aa in translated_with_gaps:\n",
    "        if aa == \"-\":\n",
    "            untranslated_with_gaps.append(\"---\")  # gap at codon level\n",
    "        else:\n",
    "            codon = untranslated_no_gaps[nt_index:nt_index+3]\n",
    "            untranslated_with_gaps.append(codon)\n",
    "            nt_index += 3\n",
    "\n",
    "    return \"\".join(untranslated_with_gaps)\n",
    "\n",
    "\n",
    "\n",
    "list_of_untranslated_files = os.listdir(f'{location}/3.Sequences_no_gap')\n",
    "\n",
    "list_of_aligned_files = os.listdir(f'{location}/5.Aligned_MAFFT_protein')\n",
    "\n",
    "list_of_folders =  os.listdir(f\"{location}\")\n",
    "if \"6.Aligned_MAFFT_untranslated\" not in list_of_folders:\n",
    "    os.mkdir(f\"{location}/6.Aligned_MAFFT_untranslated\")\n",
    "else:\n",
    "    subprocess.run(f\"rm -r {location}/6.Aligned_MAFFT_untranslated/*\", shell = True)\n",
    "\n",
    "for files in list_of_aligned_files:\n",
    "    aligned_file_dictionary = SeqIO.to_dict(SeqIO.parse(f'{location}/5.Aligned_MAFFT_protein/{files}', \"fasta\"))\n",
    "    untrans_file_dictionbary = SeqIO.to_dict(SeqIO.parse(f'{location}/3.Sequences_no_gap/{files}', \"fasta\"))\n",
    "    output = ''\n",
    "    for sequence_name in aligned_file_dictionary:\n",
    "        aligned_sequence = aligned_file_dictionary[sequence_name].seq\n",
    "        untrans_sequence =  untrans_file_dictionbary[sequence_name].seq\n",
    "        # print(aligned_sequence)\n",
    "        output += f\">{sequence_name}\\n{(map_gaps_to_untranslated(str(aligned_sequence), str(untrans_sequence)))}\\n\"\n",
    "    with open(f'{location}/6.Aligned_MAFFT_untranslated/{files}', 'w') as out_file:\n",
    "        out_file.write(output)\n",
    "\n",
    "print(\"Step 3: Done\")\n",
    "\n",
    "print(f\"Step 4: Counting species per gene and removing those with < {int(len(species_name_list)/2)} missing\")\n",
    "list_of_files = os.listdir(f'{location}/6.Aligned_MAFFT_untranslated')\n",
    "output = \"Sequence,Count\"\n",
    "for i,files in enumerate(list_of_files):\n",
    "    if i % 100 == 0:\n",
    "        print(i, \"Done\")\n",
    "    aligned_file_dictionary = SeqIO.to_dict(SeqIO.parse(f'{location}/6.Aligned_MAFFT_untranslated/{files}', \"fasta\"))\n",
    "    output += f\"\\n{files},{len(aligned_file_dictionary)}\"\n",
    "    if len(aligned_file_dictionary) < len(species_name_list)/2:\n",
    "        subprocess.run(f\"rm {location}/6.Aligned_MAFFT_untranslated/{files}\", shell = True)\n",
    "\n",
    "with open(f'{location}/counts.csv', 'w') as out_file:\n",
    "    out_file.write(output)\n",
    "\n",
    "print(\"Step 4: Done\")\n",
    "\n",
    "print(f\"Step 5: Trimming with Clipkit\")\n",
    "\n",
    "list_of_folders =  os.listdir(f\"{location}\")\n",
    "if \"7.Clipkit_trimmed\" not in list_of_folders:\n",
    "    os.mkdir(f\"{location}/7.Clipkit_trimmed\")\n",
    "else:\n",
    "    subprocess.run(f\"rm -r {location}/7.Clipkit_trimmed/*\", shell = True)\n",
    "\n",
    "\n",
    "list_of_files = os.listdir(f\"{location}/6.Aligned_MAFFT_untranslated\")\n",
    "\n",
    "def run_clipkit(file_name):\n",
    "    cmd = (\n",
    "        f\"clipkit '{location}/6.Aligned_MAFFT_untranslated/{file_name}' \"\n",
    "        f\"--codon -o '{location}/7.Clipkit_trimmed/{file_name}'\"\n",
    "    )\n",
    "    subprocess.run(cmd, shell=True)\n",
    "    return file_name\n",
    "\n",
    "with ThreadPoolExecutor(max_workers=15) as executor:\n",
    "    futures = [executor.submit(run_clipkit, f) for f in list_of_files]\n",
    "\n",
    "    for i, future in enumerate(as_completed(futures), 1):\n",
    "        print(f\"{i}/{len(futures)} completed: {future.result()}\")\n",
    "\n",
    "print(f\"Step 5: Done\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67152b00",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Step 6: Make Tree with random 200 genes\")\n",
    "\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "import subprocess\n",
    "import random\n",
    "\n",
    "location = \"/mnt/griffin/saubar/Species_sequences\" # location of the root folder where all data and analysis will be done \n",
    "\n",
    "list_of_folders =  os.listdir(f\"{location}\")\n",
    "if \"8.For_tree\" not in list_of_folders:\n",
    "    os.mkdir(f\"{location}/8.For_tree\")\n",
    "else:\n",
    "    subprocess.run(f\"rm -r {location}/8.For_tree/*\", shell = True)\n",
    "\n",
    "\n",
    "species_dict = {}\n",
    "\n",
    "with open(f'{location}/counts.csv', 'r') as count_file:\n",
    "    count_lines = count_file.readlines()[1:]\n",
    "\n",
    "max_species = 0\n",
    "for lines in count_file.readlines()[1:]:\n",
    "    lines_split = lines.strip().split(\",\")\n",
    "    if int(lines_split[1]) > max_species:\n",
    "            max_species = int(lines_split[1])\n",
    "gene_to_be_used = []\n",
    "for lines in count_file.readlines()[1:]:\n",
    "    lines_split = lines.strip().split(\",\")\n",
    "    if int(lines_split[1]) == max_species:\n",
    "           gene_to_be_used.append(lines_split[0]) \n",
    "\n",
    "count_gene = 0  \n",
    "for gene_name in   gene_to_be_used:\n",
    "                  \n",
    "    if count_gene > 200:\n",
    "        break\n",
    "    count_gene += 1\n",
    "    \n",
    "    # print(lines_split[0])\n",
    "    fasta_file = SeqIO.parse(f\"{location}/7.Clipkit_trimmed/{random.choice(gene_to_be_used)}\", \"fasta\")\n",
    "    for records in fasta_file:\n",
    "        species_dict.setdefault(records.id, '')\n",
    "        species_dict[records.id] = species_dict[records.id] + records.seq\n",
    "output = ''\n",
    "for key, value in species_dict.items():\n",
    "    output += f\">{key}\\n{value}\\n\\n\"\n",
    "\n",
    "with open(f\"{location}/8.For_tree/merged_{max_species}sp.fas\", 'w') as out_file:\n",
    "    out_file.write(output)\n",
    "\n",
    "\n",
    "subprocess.run(f\"cd '{location}/8.For_tree'\\niqtree2 -s merged_{max_species}sp.fas -m MFP -nt AUTO\", shell = True)\n",
    "\n",
    "print(f\"Step 6: Make Tree with random 200 genes\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "genomics_work",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
